File: java/src/main/java/problems1_100/RomanToInteger.java
package problems1_100;

import java.util.HashMap;
import java.util.Map;

enum RomanToInteger {
    DEFAULT {
        @Override
        int romanToInt(String s) {
            int result = 0;
            for (int i = 0; i < s.length(); i++) {
                char c = s.charAt(i);
                if (i > 0 && ROMANS.get(c) > ROMANS.get(s.charAt(i - 1))) {
                    result += ROMANS.get(c) - 2 * ROMANS.get(s.charAt(i - 1));
                } else {
                    result += ROMANS.get(c);
                }
            }
            return result;
        }
    };

    abstract int romanToInt(String s);

    private static final Map<Character, Integer> ROMANS = new HashMap<>();

    static {
        ROMANS.put('I', 1);
        ROMANS.put('V', 5);
        ROMANS.put('X', 10);
        ROMANS.put('L', 50);
        ROMANS.put('C', 100);
        ROMANS.put('D', 500);
        ROMANS.put('M', 1000);
    }
}

File: kotlin/src/main/kotlin/problems1_100/RomanToInteger.kt
package problems1_100

enum class RomanToInteger {
    DEFAULT {
        override fun romanToInt(s: String): Int {
            var result = 0
            for (i in s.indices) {
                val c = s[i]
                result +=
                    if (i > 0 && ROMANS[c]!! > ROMANS[s[i - 1]]!!) {
                        ROMANS[c]!! - 2 * ROMANS[s[i - 1]]!!
                    } else {
                        ROMANS[c]!!
                    }
            }
            return result
        }
    },
    ;

    abstract fun romanToInt(s: String): Int

    companion object {
        private val ROMANS =
            hashMapOf(
                'I' to 1,
                'V' to 5,
                'X' to 10,
                'L' to 50,
                'C' to 100,
                'D' to 500,
                'M' to 1000,
            )
    }
}
